### PR TITLE
feat(semver): Add function that checks for semver versioning

### PR DESCRIPTION
**Pull request by ahmedetefy**
This PR:
- Adds property on `Release` to identify if an instance of Release model follows semver
- Adds static method that checks if a release version follows semantic versioning.
- Adds placeholder function to be used in the future for making decisions on whether to follow semantic versioning based ordering or fallback to date based ordering. This function decided if we should follow semver or not based on if the latest 10 releases of a project follow semver and based on if an optional release version argument follows semver

### CODE REVIEW

File: tests/sentry/models/test_release.py
Comment (wedamija) [inline tests/sentry/models/test_release.py:56]: Could we add a case here that includes a number > 63 bits long?

File: src/sentry/models/release.py
Comment (wedamija) [inline src/sentry/models/release.py:372]: If you want to know whether a release follows semver, you can just check `release.package is not None`. If any semver field is filled, the release is semver. I don't think we need to rely on reparsing the version here.

I think this could also just be a property on the release rather than a `@staticmethod`. 

However, if you want this method to tell whether a project is using semver, it might be more useful to do `Release.objects.filter_to_semver().filter(organization=<org_id>, projects__in=[myproject]).exists()` (might be useful to filter to releases created in the last `<period of time>`)

Comment (ahmedetefy) [reply src/sentry/models/release.py:372]: I agree with you we should probably have `is_semver` property so that we know if a release object follows semver but this function serves a different purpose; Its a static method that takes a release string (not a release object) and returns whether that release string follows semver

Comment (ahmedetefy) [reply src/sentry/models/release.py:376]: Added the property + the changes we discussed on Friday in regards to checking for the latest ten releases if the follow semver or not

Comment (wedamija) [reply src/sentry/models/release.py:376]: Right, I think I left this comment before we discussed via hangouts. It might make sense to rename `release` to `version` to make it clearer

File: src/sentry/models/release.py
Comment (ahmedetefy) [reply src/sentry/models/release.py:988]: I am skeptical about `any` because what if only one coincidentally matches semver maybe we can change it to atleast like 7/10 of them have to be semver?

Comment (wedamija) [reply src/sentry/models/release.py:988]: It's kind of tricky because there are a few cases we will handle here:
- User just started using semver, ideally we want to start marking them as using semver fairly quickly
- User made a mistake and accidentally stopped using semver a couple times
- User made a mistake and accidentally started using semver a couple times
- User stopped using semver

What if we checked that there is a semver release in the most recent 2-3 releases, as well as having at least 3(?) releases using semver in the last 10?

Comment (ahmedetefy) [reply src/sentry/models/release.py:988]: I think that's a good idea 
What about if the user has less than 10 releases? How would we check for that?

Comment (ahmedetefy) [reply src/sentry/models/release.py:988]: Maybe in that case we forego the next condition i.e. forego the atleast 3 releases in the last 10 and finding one semver in the last 3 should suffice.. wdyt

Comment (wedamija) [reply src/sentry/models/release.py:988]: Yeah, that sounds good to me!

**Review by wedamija — APPROVED** [review]
Approving so that you're unblocked tomorrow, but make sure to cover the `.distinct` case before merging

File: src/sentry/models/release.py
Comment (ahmedetefy) [reply src/sentry/models/release.py:985]: How would that happen though If I am passing only one `project_id` at a time?

Comment (wedamija) [reply src/sentry/models/release.py:1069]: Good point, I'm just used to seeing joins like this and warning about `.distinct`

**Pull request merged**
